src/main/scala/is/hail/utils/richUtils/RichIndexedRowMatrix.scala

```diff
@@ -21,140 +26,57 @@ object RichIndexedRowMatrix {
         s"rowsPerBlock needs to be greater than 0. rowsPerBlock: $rowsPerBlock")
       require(colsPerBlock > 0,
         s"colsPerBlock needs to be greater than 0. colsPerBlock: $colsPerBlock")
-      val m = indexedRowMatrix.numRows().toInt
-      val n = indexedRowMatrix.numCols().toInt
 
-      val blocksRDD: RDD[((Int, Int), Matrix)] = indexedRowMatrix.rows.flatMap({ ir =>
+      val m = indexedRowMatrix.numRows()
+      val n = indexedRowMatrix.numCols()
+      val lastRowBlockIndex = m / rowsPerBlock
+      val lastColBlockIndex = n / colsPerBlock
+      val lastRowBlockSize = (m % rowsPerBlock).toInt
+      val lastColBlockSize = (n % colsPerBlock).toInt
+      val numRowBlocks = math.ceil(m.toDouble / rowsPerBlock).toInt
+      val numColBlocks = math.ceil(n.toDouble / colsPerBlock).toInt
+
+      val intClass = classTag[Int].runtimeClass
+      val gpConstructor = Class.forName("org.apache.spark.mllib.linalg.distributed.GridPartitioner")
+        .getDeclaredConstructor(intClass,intClass,intClass,intClass)
+      def sneakyGridPartitioner(nRowBlocks: Int, nColBlocks: Int, nRowsPerPart: Int, nColsPerPart: Int): Partitioner = try {
+        gpConstructor.setAccessible(true)
+        gpConstructor.newInstance(nRowBlocks: java.lang.Integer, nColBlocks: java.lang.Integer,
+          nRowsPerPart: java.lang.Integer, nColsPerPart: java.lang.Integer).asInstanceOf[Partitioner]
+      } finally {
+        gpConstructor.setAccessible(false)
+      }
```

Comment:
We should lift these reflection related values and def onto the object since they have no dependence on `rowsPerBlock` or `colsPerBlock`